File: work/claude-code/negaboku/Unity/Scripts/Systems/Battle/BattleSystem.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;
using NegabokuRPG.Data;
using NegabokuRPG.Characters;

namespace NegabokuRPG.Systems
{
    /// <summary>
    /// バトルシステム - ターン制戦闘の管理
    /// </summary>
    public class BattleSystem : MonoBehaviour
    {
        [Header("バトル設定")]
        [SerializeField] private float actionAnimationDuration = 1.0f;
        [SerializeField] private float turnDelay = 0.5f;

        // バトル状態
        private List<BattleCharacter> playerParty = new List<BattleCharacter>();
        private List<BattleEnemy> enemies = new List<BattleEnemy>();
        private List<BattleCombatant> turnOrder = new List<BattleCombatant>();
        private int currentTurnIndex = 0;
        private bool battleActive = false;

        // 結果
        private BattleResult battleResult;

        // イベント
        public event Action OnBattleStart;
        public event Action<BattleResult> OnBattleEnd;
        public event Action<BattleCombatant> OnTurnStart;
        public event Action<BattleAction> OnActionExecuted;
        public event Action<BattleCombatant, int> OnDamageDealt;
        public event Action<BattleCombatant, int> OnHealingDone;

        // 依存システム
        private RelationshipSystem relationshipSystem;

        private void Awake()
        {
            relationshipSystem = RelationshipSystem.Instance;
        }

        /// <summary>
        /// バトル開始
        /// </summary>
        public void StartBattle(List<PlayerCharacter> party, List<BattleEnemyData> enemyData)
        {
            InitializeBattle(party, enemyData);
            StartCoroutine(BattleLoop());
        }

        /// <summary>
        /// バトル初期化
        /// </summary>
        private void InitializeBattle(List<PlayerCharacter> party, List<BattleEnemyData> enemyData)
        {
            // プレイヤーパーティの初期化
            playerParty.Clear();
            foreach (var player in party)
            {
                var battleChar = new BattleCharacter(player);
                playerParty.Add(battleChar);
            }

            // 敵の初期化
            enemies.Clear();
            foreach (var data in enemyData)
            {
                var enemy = new BattleEnemy(data);
                enemies.Add(enemy);
            }

            // ターン順序の決定
            InitializeTurnOrder();

            battleActive = true;
            OnBattleStart?.Invoke();
        }

        /// <summary>
        /// ターン順序の初期化
        /// </summary>
        private void InitializeTurnOrder()
        {
            turnOrder.Clear();
            
            // 全戦闘参加者を追加
            turnOrder.AddRange(playerParty.Cast<BattleCombatant>());
            turnOrder.AddRange(enemies.Cast<BattleCombatant>());

            // 速度順でソート
            turnOrder = turnOrder.OrderByDescending(c => c.Speed).ToList();
            currentTurnIndex = 0;
        }

        /// <summary>
        /// バトルメインループ
        /// </summary>
        private IEnumerator BattleLoop()
        {
            while (battleActive)
            {
                var currentCombatant = turnOrder[currentTurnIndex];

                // 戦闘不能者はスキップ
                if (currentCombatant.IsDefeated)
                {
                    NextTurn();
                    continue;
                }

                // ターン開始通知
                OnTurnStart?.Invoke(currentCombatant);

                // ステータス効果処理
                currentCombatant.ProcessStatusEffects();

                // 行動実行
                if (!currentCombatant.IsDefeated)
                {
                    yield return StartCoroutine(ExecuteTurn(currentCombatant));
                }

                // バトル終了判定
                if (IsBattleOver())
                {
                    yield return StartCoroutine(EndBattle());
                    break;
                }

                yield return new WaitForSeconds(turnDelay);
                NextTurn();
            }
        }

        /// <summary>
        /// 1ターンの実行
        /// </summary>
        private IEnumerator ExecuteTurn(BattleCombatant combatant)
        {
            BattleAction action = null;

            if (combatant is BattleCharacter playerChar)
            {
                // プレイヤーキャラクターの行動選択（UI待ち）
                action = yield return StartCoroutine(WaitForPlayerAction(playerChar));
            }
            else if (combatant is BattleEnemy enemy)
            {
                // 敵の行動決定（AI）
                action = DetermineEnemyAction(enemy);
            }

            if (action != null)
            {
                yield return StartCoroutine(ExecuteAction(action));
            }
        }

        /// <summary>
        /// プレイヤーの行動選択を待機
        /// </summary>
        private IEnumerator WaitForPlayerAction(BattleCharacter character)
        {
            // UIシステムからの入力待ち
            BattleAction selectedAction = null;
            bool actionSelected = false;

            // UIイベントリスナー設定
            System.Action<BattleAction> onActionSelected = (action) => {
                selectedAction = action;
                actionSelected = true;
            };

            // UI通知（実際の実装ではUIManagerを通して行う）
            NotifyUIForActionSelection(character, onActionSelected);

            // 行動選択まで待機
            yield return new WaitUntil(() => actionSelected);

            yield return selectedAction;
        }

        /// <summary>
        /// 敵の行動決定
        /// </summary>
        private BattleAction DetermineEnemyAction(BattleEnemy enemy)
        {
            // 簡単なAI：ランダムで通常攻撃かスキル使用
            var availableSkills = enemy.GetUsableSkills();
            
            if (availableSkills.Count > 0 && UnityEngine.Random.value < 0.3f)
            {
                var skill = availableSkills[UnityEngine.Random.Range(0, availableSkills.Count)];
                var targets = SelectTargetsForSkill(skill, false);
                
                return new BattleAction
                {
                    actor = enemy,
                    actionType = BattleActionType.Skill,
                    skill = skill,
                    targets = targets
                };
            }
            else
            {
                // 通常攻撃
                var target = SelectRandomTarget(true); // プレイヤーをターゲット
                return new BattleAction
                {
                    actor = enemy,
                    actionType = BattleActionType.Attack,
                    targets = new List<BattleCombatant> { target }
                };
            }
        }

        /// <summary>
        /// 行動実行
        /// </summary>
        private IEnumerator ExecuteAction(BattleAction action)
        {
            OnActionExecuted?.Invoke(action);

            switch (action.actionType)
            {
                case BattleActionType.Attack:
                    yield return StartCoroutine(ExecuteAttack(action));
                    break;
                case BattleActionType.Skill:
                    yield return StartCoroutine(ExecuteSkill(action));
                    break;
                case BattleActionType.Item:
                    yield return StartCoroutine(ExecuteItem(action));
                    break;
                case BattleActionType.Defend:
                    yield return StartCoroutine(ExecuteDefend(action));
                    break;
            }

            yield return new WaitForSeconds(actionAnimationDuration);
        }

        /// <summary>
        /// 通常攻撃実行
        /// </summary>
        private IEnumerator ExecuteAttack(BattleAction action)
        {
            if (action.targets.Count > 0)
            {
                var target = action.targets[0];
                int damage = CalculatePhysicalDamage(action.actor, target);
                
                ApplyDamage(target, damage);
                OnDamageDealt?.Invoke(target, damage);
            }
            yield return null;
        }

        /// <summary>
        /// スキル実行
        /// </summary>
        private IEnumerator ExecuteSkill(BattleAction action)
        {
            var skill = action.skill;
            var caster = action.actor;

            // MP消費
            caster.ModifyMP(-skill.mpCost);

            // 関係値による威力修正
            float powerModifier = 1.0f;
            if (skill.skillType != SkillType.Normal && relationshipSystem != null)
            {
                var relationships = relationshipSystem.GetAllRelationships();
                powerModifier = skill.CalculatePowerModifier(relationships);
            }

            // ターゲットに効果適用
            foreach (var target in action.targets)
            {
                int effectValue = Mathf.RoundToInt(skill.basePower * powerModifier);

                if (skill.basePower > 0) // ダメージ
                {
                    ApplyDamage(target, effectValue);
                    OnDamageDealt?.Invoke(target, effectValue);
                }
                else if (skill.basePower < 0) // 回復
                {
                    int healing = Mathf.Abs(effectValue);
                    ApplyHealing(target, healing);
                    OnHealingDone?.Invoke(target, healing);
                }

                // ステータス効果適用
                foreach (var statusEffect in skill.statusEffects)
                {
                    target.AddStatusEffect(statusEffect.effectType, statusEffect.value, statusEffect.duration);
                }
            }

            // 関係値変化処理
            HandleSkillRelationshipEffects(skill, caster);

            yield return null;
        }

        /// <summary>
        /// アイテム使用実行
        /// </summary>
        private IEnumerator ExecuteItem(BattleAction action)
        {
            // アイテム効果の実装
            yield return null;
        }

        /// <summary>
        /// 防御実行
        /// </summary>
        private IEnumerator ExecuteDefend(BattleAction action)
        {
            // 防御バフ付与
            action.actor.AddStatusEffect(StatusEffectType.DefenseUp, 
                Mathf.RoundToInt(action.actor.Defense * 0.5f), 1);
            yield return null;
        }

        /// <summary>
        /// 物理ダメージ計算
        /// </summary>
        private int CalculatePhysicalDamage(BattleCombatant attacker, BattleCombatant target)
        {
            int baseDamage = attacker.Attack - target.Defense;
            baseDamage = Mathf.Max(1, baseDamage); // 最低1ダメージ

            // クリティカル判定
            if (UnityEngine.Random.value < 0.1f) // 10%でクリティカル
            {
                baseDamage = Mathf.RoundToInt(baseDamage * 1.5f);
            }

            return baseDamage;
        }

        /// <summary>
        /// ダメージ適用
        /// </summary>
        private void ApplyDamage(BattleCombatant target, int damage)
        {
            target.ModifyHP(-damage);
        }

        /// <summary>
        /// 回復適用
        /// </summary>
        private void ApplyHealing(BattleCombatant target, int healing)
        {
            target.ModifyHP(healing);
        }

        /// <summary>
        /// スキルによる関係値変化を処理
        /// </summary>
        private void HandleSkillRelationshipEffects(SkillData skill, BattleCombatant caster)
        {
            if (relationshipSystem == null || skill.requiredCharacterIds.Count < 2) return;

            string casterId = caster.CharacterId;
            
            // 共闘技の場合、関係値向上
            if (skill.skillType == SkillType.Cooperation)
            {
                foreach (var partnerId in skill.requiredCharacterIds)
                {
                    if (partnerId != casterId)
                    {
                        relationshipSystem.HandleBattleEvent(BattleEventType.Cooperation, casterId, partnerId);
                    }
                }
            }
            // 対立技の場合、関係値悪化
            else if (skill.skillType == SkillType.Conflict)
            {
                foreach (var rivalId in skill.requiredCharacterIds)
                {
                    if (rivalId != casterId)
                    {
                        relationshipSystem.HandleBattleEvent(BattleEventType.Rivalry, casterId, rivalId);
                    }
                }
            }
        }

        /// <summary>
        /// スキル用ターゲット選択
        /// </summary>
        private List<BattleCombatant> SelectTargetsForSkill(SkillData skill, bool isPlayerAction)
        {
            var targets = new List<BattleCombatant>();

            switch (skill.targetType)
            {
                case TargetType.Self:
                    // 実装時はキャスターを返す
                    break;
                case TargetType.SingleEnemy:
                    if (isPlayerAction)
                        targets.Add(SelectRandomTarget(false)); // 敵をターゲット
                    else
                        targets.Add(SelectRandomTarget(true));  // プレイヤーをターゲット
                    break;
                case TargetType.AllEnemies:
                    if (isPlayerAction)
                        targets.AddRange(enemies.Where(e => !e.IsDefeated).Cast<BattleCombatant>());
                    else
                        targets.AddRange(playerParty.Where(p => !p.IsDefeated).Cast<BattleCombatant>());
                    break;
                // 他のターゲットタイプも実装
            }

            return targets;
        }

        /// <summary>
        /// ランダムターゲット選択
        /// </summary>
        private BattleCombatant SelectRandomTarget(bool selectPlayer)
        {
            if (selectPlayer)
            {
                var alivePlayers = playerParty.Where(p => !p.IsDefeated).ToList();
                return alivePlayers.Count > 0 ? alivePlayers[UnityEngine.Random.Range(0, alivePlayers.Count)] : null;
            }
            else
            {
                var aliveEnemies = enemies.Where(e => !e.IsDefeated).ToList();
                return aliveEnemies.Count > 0 ? aliveEnemies[UnityEngine.Random.Range(0, aliveEnemies.Count)] : null;
            }
        }

        /// <summary>
        /// バトル終了判定
        /// </summary>
        private bool IsBattleOver()
        {
            bool allPlayersDefeated = playerParty.All(p => p.IsDefeated);
            bool allEnemiesDefeated = enemies.All(e => e.IsDefeated);

            return allPlayersDefeated || allEnemiesDefeated;
        }

        /// <summary>
        /// バトル終了処理
        /// </summary>
        private IEnumerator EndBattle()
        {
            battleActive = false;

            bool playerVictory = enemies.All(e => e.IsDefeated);
            
            battleResult = new BattleResult
            {
                victory = playerVictory,
                experience = playerVictory ? enemies.Sum(e => e.ExpReward) : 0,
                gold = playerVictory ? enemies.Sum(e => e.GoldReward) : 0,
                items = playerVictory ? enemies.SelectMany(e => e.ItemRewards).ToList() : new List<string>()
            };

            OnBattleEnd?.Invoke(battleResult);
            yield return null;
        }

        /// <summary>
        /// 次のターンへ進む
        /// </summary>
        private void NextTurn()
        {
            currentTurnIndex = (currentTurnIndex + 1) % turnOrder.Count;
        }

        /// <summary>
        /// UI通知（実際の実装ではUIManagerを使用）
        /// </summary>
        private void NotifyUIForActionSelection(BattleCharacter character, System.Action<BattleAction> callback)
        {
            // UIシステムへの通知処理
            // 実際の実装では、UIManagerを通してUI表示し、プレイヤーの選択を待つ
        }

        /// <summary>
        /// 外部からの行動入力受付
        /// </summary>
        public void InputPlayerAction(BattleAction action)
        {
            // UIからの行動入力を受け取る
            // 実際の実装では、現在の行動選択待ちキャラクターと照合してから処理
        }
    }

    /// <summary>
    /// バトル行動データ
    /// </summary>
    [Serializable]
    public class BattleAction
    {
        public BattleCombatant actor;
        public BattleActionType actionType;
        public SkillData skill;
        public string itemId;
        public List<BattleCombatant> targets;
    }

    /// <summary>
    /// バトル行動タイプ
    /// </summary>
    public enum BattleActionType
    {
        Attack,
        Skill,
        Item,
        Defend,
        Escape
    }

    /// <summary>
    /// バトル結果
    /// </summary>
    [Serializable]
    public class BattleResult
    {
        public bool victory;
        public int experience;
        public int gold;
        public List<string> items;
    }
}

File: work/claude-code/negaboku/src/systems/BattleSystem.ts
import { Character, Skill, SkillType } from '../types/Character';
import { BattleCharacter, BattleAction, BattleActionType, BattleResult, Enemy, StatusEffect, StatusEffectType } from '../types/Battle';
import { RelationshipSystem, BattleEventType } from './RelationshipSystem';

export class BattleSystem {
  private playerParty: BattleCharacter[];
  private enemies: Enemy[];
  private relationshipSystem: RelationshipSystem;
  private turnOrder: string[];
  private currentTurn: number;

  constructor(party: Character[], enemies: Enemy[], relationshipSystem: RelationshipSystem) {
    this.playerParty = party.map(char => this.characterToBattleCharacter(char));
    this.enemies = enemies.map(enemy => this.initializeEnemy(enemy));
    this.relationshipSystem = relationshipSystem;
    this.turnOrder = [];
    this.currentTurn = 0;
    this.initializeTurnOrder();
  }

  private initializeEnemy(enemy: Enemy): Enemy {
    return {
      ...enemy,
      currentHp: enemy.hp,
      statusEffects: []
    };
  }

  private characterToBattleCharacter(character: Character): BattleCharacter {
    return {
      ...character,
      currentHp: character.hp,
      currentMp: character.mp,
      statusEffects: [],
      position: 0
    };
  }

  private initializeTurnOrder(): void {
    const allCombatants = [
      ...this.playerParty.map(char => ({ id: char.id, speed: char.speed, type: 'player' })),
      ...this.enemies.map(enemy => ({ id: enemy.id, speed: enemy.speed, type: 'enemy' }))
    ];

    this.turnOrder = allCombatants
      .sort((a, b) => b.speed - a.speed)
      .map(combatant => combatant.id);
  }

  executeAction(action: BattleAction): void {
    const actor = this.getCombatantById(action.actorId);
    if (!actor || this.isDefeated(actor)) return;

    switch (action.type) {
      case BattleActionType.ATTACK:
        this.executeBasicAttack(action);
        break;
      case BattleActionType.SKILL:
        this.executeSkill(action);
        break;
      case BattleActionType.DEFEND:
        this.executeDefend(action);
        break;
    }

    this.processStatusEffects();
    this.nextTurn();
  }

  private executeBasicAttack(action: BattleAction): void {
    const attacker = this.getCombatantById(action.actorId);
    const target = this.getCombatantById(action.targetIds[0]);
    
    if (!attacker || !target) return;

    const damage = Math.max(1, attacker.attack - target.defense);
    this.dealDamage(target, damage);
  }

  private executeSkill(action: BattleAction): void {
    const caster = this.getCombatantById(action.actorId);
    if (!caster || !action.skillId) return;

    const skill = this.getSkillById(caster, action.skillId);
    if (!skill) return;

    const currentMp = 'currentMp' in caster ? caster.currentMp : 0;
    if (currentMp < skill.mpCost) return;

    if ('currentMp' in caster) {
      caster.currentMp -= skill.mpCost;
    }

    switch (skill.type) {
      case SkillType.NORMAL:
        this.executeNormalSkill(caster, skill, action.targetIds);
        break;
      case SkillType.COOPERATION:
        this.executeCooperationSkill(caster, skill, action.targetIds);
        break;
      case SkillType.CONFLICT:
        this.executeConflictSkill(caster, skill, action.targetIds);
        break;
    }
  }

  private executeNormalSkill(caster: BattleCharacter | Enemy, skill: Skill, targetIds: string[]): void {
    for (const targetId of targetIds) {
      const target = this.getCombatantById(targetId);
      if (!target) continue;

      if (skill.power > 0) {
        const damage = skill.power;
        this.dealDamage(target, damage);
      } else if (skill.power < 0) {
        const healing = Math.abs(skill.power);
        this.heal(target, healing);
      }
    }
  }

  private executeCooperationSkill(caster: BattleCharacter | Enemy, skill: Skill, targetIds: string[]): void {
    const partnerId = skill.requiredCharacters?.[0];
    if (partnerId && this.relationshipSystem.canUseCooperationSkill(caster.id, partnerId)) {
      const enhancedPower = skill.power * 1.5;
      
      for (const targetId of targetIds) {
        const target = this.getCombatantById(targetId);
        if (target) {
          this.dealDamage(target, enhancedPower);
        }
      }

      this.relationshipSystem.handleBattleEvent(BattleEventType.COOPERATION, caster.id, partnerId);
    }
  }

  private executeConflictSkill(caster: BattleCharacter | Enemy, skill: Skill, targetIds: string[]): void {
    const rivalId = skill.requiredCharacters?.[0];
    if (rivalId && this.relationshipSystem.canUseConflictSkill(caster.id, rivalId)) {
      const enhancedPower = skill.power * 1.3;
      
      for (const targetId of targetIds) {
        const target = this.getCombatantById(targetId);
        if (target) {
          this.dealDamage(target, enhancedPower);
        }
      }

      this.relationshipSystem.handleBattleEvent(BattleEventType.RIVALRY, caster.id, rivalId);
    }
  }

  private executeDefend(action: BattleAction): void {
    const defender = this.getCombatantById(action.actorId);
    if (!defender) return;

    const defenseBoost: StatusEffect = {
      id: 'defense_boost',
      name: '防御',
      duration: 1,
      type: StatusEffectType.DEFENSE_UP,
      value: Math.floor(defender.defense * 0.5)
    };

    defender.statusEffects.push(defenseBoost);
  }

  private dealDamage(target: BattleCharacter | Enemy, damage: number): void {
    const finalDamage = Math.max(1, damage);
    target.currentHp = Math.max(0, target.currentHp - finalDamage);
  }

  private heal(target: BattleCharacter | Enemy, healing: number): void {
    let maxHp: number;
    if ('maxHp' in target && target.maxHp !== undefined) {
      maxHp = target.maxHp;
    } else if ('hp' in target) {
      maxHp = target.hp;
    } else {
      maxHp = 100; // デフォルト値
    }
    target.currentHp = Math.min(maxHp, target.currentHp + healing);
  }

  private processStatusEffects(): void {
    const allCombatants: (BattleCharacter | Enemy)[] = [...this.playerParty, ...this.enemies];
    
    for (const combatant of allCombatants) {
      combatant.statusEffects = combatant.statusEffects.filter((effect: StatusEffect) => {
        this.applyStatusEffect(combatant, effect);
        effect.duration--;
        return effect.duration > 0;
      });
    }
  }

  private applyStatusEffect(combatant: BattleCharacter | Enemy, effect: StatusEffect): void {
    switch (effect.type) {
      case StatusEffectType.POISON:
        this.dealDamage(combatant, effect.value);
        break;
    }
  }

  private getCombatantById(id: string): BattleCharacter | Enemy | undefined {
    return this.playerParty.find(char => char.id === id) || 
           this.enemies.find(enemy => enemy.id === id);
  }

  private getSkillById(combatant: BattleCharacter | Enemy, skillId: string): Skill | undefined {
    if ('skills' in combatant && Array.isArray(combatant.skills)) {
      return (combatant.skills as Skill[]).find(skill => skill.id === skillId);
    }
    return undefined;
  }

  private isDefeated(combatant: BattleCharacter | Enemy): boolean {
    return combatant.currentHp <= 0;
  }

  private nextTurn(): void {
    this.currentTurn = (this.currentTurn + 1) % this.turnOrder.length;
  }

  isBattleOver(): boolean {
    const playersAlive = this.playerParty.some(char => !this.isDefeated(char));
    const enemiesAlive = this.enemies.some(enemy => !this.isDefeated(enemy));
    
    return !playersAlive || !enemiesAlive;
  }

  getBattleResult(): BattleResult {
    const playersWon = this.enemies.every(enemy => this.isDefeated(enemy));
    
    return {
      winner: playersWon ? 'player' : 'enemy',
      experience: playersWon ? this.enemies.reduce((sum, enemy) => sum + enemy.experience, 0) : 0,
      gold: playersWon ? this.enemies.reduce((sum, enemy) => sum + enemy.gold, 0) : 0,
      items: playersWon ? this.enemies.flatMap(enemy => enemy.dropItems) : [],
      relationshipChanges: this.relationshipSystem.getAllRelationships()
    };
  }

  getCurrentTurnActor(): string {
    return this.turnOrder[this.currentTurn];
  }

  getPlayerParty(): BattleCharacter[] {
    return [...this.playerParty];
  }

  getEnemies(): Enemy[] {
    return [...this.enemies];
  }
}